### ISSUE TITLE
memory_limit setting doesn't work

### ISSUE THREAD

**Issue by michaelmdresser** [duckdb/duckdb#8398]
### What happens?

With `memory_limit` set to 50MB, DuckDB uses ~200MiB of memory according to `/usr/bin/time -l`. I have observed similar behavior where DuckDB exceeds higher memory limits (measured in GB) but this is a minimal reproduction. The output of the repro command is this:

```
Inserted row total: 100000
Inserted row total: 200000
Inserted row total: 300000
Inserted row total: 400000
Inserted row total: 500000
┌─────────────────────────────────┐
│ current_setting('memory_limit') │
│             varchar             │
├─────────────────────────────────┤
│ 100.0MB                         │
└─────────────────────────────────┘
        0.29 real         0.47 user         0.12 sys
           216334336  maximum resident set size
                   0  average shared memory size
                   0  average unshared data size
                   0  average unshared stack size
               13314  page reclaims
                  30  page faults
                   0  swaps
                   0  block input operations
                   0  block output operations
                   0  messages sent
                   0  messages received
                   0  signals received
                 205  voluntary context switches
                 902  involuntary context switches
          5424799228  instructions retired
          1842084519  cycles elapsed
           194168320  peak memory footprint
```

Note that 216334336 is 206.3125 MiB.

When I reattempted with DuckDB version `v0.8.2-dev2133 726b6d1566`, I got a reported `214843392` max res set size, equal to 204.890625 MiB

### To Reproduce

Create a file `ddbdata.py` which writes data into a DuckDB file:
```py
import random
import string
import duckdb

con = duckdb.connect('memtest.duckdb')
con.sql('CREATE TABLE test (v1 VARCHAR, v2 VARCHAR)')
n_rows = 500_000
strlen = 50
for i in range(1, n_rows  + 1):
    randstr1 = ''.join(random.choices(string.ascii_uppercase, k=strlen))
    randstr2 = ''.join(random.choices(string.ascii_uppercase, k=strlen))
    con.execute('INSERT INTO test VALUES (?, ?)', (randstr1, randstr2))
    if i % 100_000 == 0:
        print("Inserted row total:", i)
con.close()
```

Create a file `duckmem.sql` with the following queries:
```sql
SET memory_limit='100MB';
SELECT current_setting('memory_limit');
CREATE SEQUENCE serial START 101;
CREATE TABLE foo AS (SELECT CONCAT(v1, v2), random(), nextval('serial')  FROM test);
```

Run the following command:
`python3 ddbdata.py && /usr/bin/time -l duckdb memtest.duckdb -c '.read duckmem.sql'`

### OS:

Darwin 22.6.0 arm64

### DuckDB Version:

v0.8.1 6536a77232

### DuckDB Client:

CLI

### Full Name:

Michael Dresser

### Affiliation:

Stackwatch (Kubecost)

### Have you tried this on the latest `master` branch?

- [X] I agree

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] I agree

**Comment by michaelmdresser** [comment]
I think this might be related to https://github.com/duckdb/duckdb/issues/8397

**Comment by Tishj** [comment]
> I think this might be related to #8397
> 
> cc @ameijer

They are not related.
Memory limit is configurable after database start, unlike access_mode

**Comment by michaelmdresser** [comment]
Understood, thank you. Is it possible I am using the `memory_limit` parameter incorrectly?

**Comment by hawkfish** [comment]
I have seen this a lot when stress testing window and other operators. I think what is happening is that the memory limit only applies buffer pool memory?

**Comment by michaelmdresser** [comment]
@hawkfish is this expected behavior when doing large aggregations/generally complex ops? To exceed the memory limit? I'm not versed in DuckDB's implementation, so I'm not sure what the implication of "the memory limit only applies buffer pool memory" is.

(I love DuckDB so far, by the way!)

**Comment by hawkfish** [comment]
Yes. If you think about it, there are situations where you need to keep a lot of data around, even if you are not using it all at once. Making large aggregations or window partitions are not situations where you can look at data once and move on.

The bit about "buffer pool memory" means that when we are doing operations where there are large chunks of data we have to keep around that might not fit in memory, we put it on large memory blocks of about the same size ("buffers") that  can be written to disk when they are not needed. These are managed by a buffer pool and it is easy to keep track of those large blocks and that is what the memory limit applies to. But there are situations where we keep everything in memory (like large window partitions) and that is not counted (although maybe it should be).

**Comment by michaelmdresser** [comment]
> there are situations where you need to keep a lot of data around, even if you are not using it all at once. Making large aggregations or window partitions are not situations where you can look at data once and move on.

This makes sense.

> The bit about "buffer pool memory" means that when we are doing operations where there are large chunks of data we have to keep around that might not fit in memory, we put it on large memory blocks of about the same size ("buffers") that can be written to disk when they are not needed. These are managed by a buffer pool and it is easy to keep track of those large blocks and that is what the memory limit applies to.

Thank you for the education!

> But there are situations where we keep everything in memory (like large window partitions) and that is not counted (although maybe it should be).

Intuitively I expected a blanket "memory limit" parameter to cap all memory usage, perhaps forcing some sort of expensive manual page-to-disk behavior in the worst case. I'd bet implementing something like that is challenging and results in questionable gains (prompting a counterargument: why not just use swap space for that behavior).

I'd be happy to contribute to some documentation here, but I doubt that I have the necessary knowledge of DuckDB's memory usage fundamentals.

**Comment by michaelmdresser** [comment]
For closure: in dev work on a project using DuckDB, I've finally noticed the `MEMORY_LIMIT` parameter kicking in. For me, that pretty much confirms that my situation is as suggested: mostly running queries that can't have intermediate data put in the "buffers" and thus mostly not running into the memory limit in DuckDB.

**Comment by szarnyasg** [comment]
At the moment, this is expected behaviour, not everything in DuckDB respects the memory limit yet. In the future, this will change.

**Comment by szarnyasg** [comment]
There's now more detailed documentation on this at https://duckdb.org/docs/current/configuration/pragmas#memory-limit
